servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.ConnectionFactory;
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.client.api.LoadBalancer;
+import io.servicetalk.client.api.LoadBalancerFactory;
+import io.servicetalk.client.api.ServiceDiscovererEvent;
+import io.servicetalk.concurrent.api.Publisher;
+
+import java.util.function.Predicate;
+
+/**
+ * {@link LoadBalancerFactory} that creates {@link LoadBalancer} instances which use a round robin strategy
+ * for selecting addresses. The created instances have the following behaviour:
+ * <ul>
+ * <li>Round robining is done at address level.</li>
+ * <li>Connections are created lazily, without any concurrency control on their creation.
+ * This can lead to over-provisioning connections when dealing with a requests surge.</li>
+ * <li>Existing connections are reused unless a selector passed to {@link LoadBalancer#selectConnection(Predicate)}
+ * suggests otherwise. This can lead to situations where connections will be used to their maximum capacity
+ * (for example in the context of pipelining) before new connections are created.</li>
+ * <li>Closed connections are automatically pruned.</li>
+ * <li>By default, connections to addresses marked as {@link ServiceDiscovererEvent#isAvailable() unavailable}
+ * are used for requests, but no new connections are created for them. In case the address' connections are busy,
+ * another host is tried. If all hosts are busy, selection fails with a
+ * {@link io.servicetalk.client.api.ConnectionRejectedException}. If {@link #eagerConnectionShutdown} is set to true,
+ * connections are immediately closed for an {@link ServiceDiscovererEvent#isAvailable() unavailable} address.</li>
+ * </ul>
+ *
+ * @param <ResolvedAddress> The resolved address type.
+ * @param <C> The type of connection.
+ */
+public final class RoundRobinLoadBalancerFactory<ResolvedAddress, C extends LoadBalancedConnection>
+        implements LoadBalancerFactory<ResolvedAddress, C> {
+
+    private final boolean eagerConnectionShutdown;
+
+    private RoundRobinLoadBalancerFactory(boolean eagerConnectionShutdown) {
+        this.eagerConnectionShutdown = eagerConnectionShutdown;
+    }
+
+    @Override
+    public <T extends C> LoadBalancer<T> newLoadBalancer(
+            final Publisher<? extends ServiceDiscovererEvent<ResolvedAddress>> eventPublisher,
+            final ConnectionFactory<ResolvedAddress, T> connectionFactory) {
+        return new RoundRobinLoadBalancer<>(eventPublisher, connectionFactory, eagerConnectionShutdown);
+    }
+
+    /**
+     * Builder for {@link RoundRobinLoadBalancerFactory}.
+     *
+     * @param <ResolvedAddress> The resolved address type.
+     * @param <C> The type of connection.
+     */
+    public static final class Builder<ResolvedAddress, C extends LoadBalancedConnection> {
+        private boolean eagerConnectionShutdown;
```

Comment:
I'm a bit hesitant of enforcing the new behavior by default without proper validation. WDYT if we make it opt-in in the next release, validate with some users, and if everything works as expected we can make it opt-out in the following releases.

Reply:
Agreed, my original idea was to introduce it as an opt-in. I committed the change.